File: Algorithms on Graphs/Assignment1/Reachability.java
import java.util.ArrayList;
import java.util.Scanner;

public class Reachability {
    private ArrayList<Integer>[] adj;
    private ArrayList<Integer> reachable;
    
    public Reachability(ArrayList<Integer>[] adj_) {
        adj = adj_;
        reachable = new ArrayList<>();
    }
    
    private boolean find(ArrayList<Integer> arr, int v) {
        for (int i = 0; i < arr.size(); i++) {
            if (v == arr.get(i)) {
                return true;
            }
        }
        return false;
    }
    
    private void DFS(int i) {
        reachable.add(i);
        for (int v: adj[i]) {
            if (find(reachable, v)) {
                continue;
            }
            DFS(v);
        }
    }
    private int reach(int x, int y) {
        reachable.clear();
        DFS(x);
        if (find(reachable, y)) {
            return 1;
        }
        return 0;
    }
    
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        Reachability reach_ = new Reachability(adj);
        System.out.println(reach_.reach(x, y));
    }
}

File: Algorithms on Graphs/Assignment1/reachability.cpp
#include <iostream>
#include <vector>

using namespace std;

class Reachability {
private:
    vector<vector<int> > adj;
    vector<int> reachable;
    
    bool find(vector<int>& vec, int x) {
        for (int i = 0; i < vec.size(); i++) {
            if (x == vec[i]) {
                return true;
            }
        }
        return false;
    }
    
    void DFS(int i) {
        reachable.push_back(i);
        for (auto v: adj[i]) {
            if (find(reachable, v)) {
                continue;
            }
            else {
                DFS(v);
            }
        }
    }
    
public:
    Reachability(vector<vector<int> > &adjcency): adj(adjcency) {}
    
    int reach(int x, int y) {
        reachable.clear();
        DFS(x);
        if (find(reachable, y)) {
            return 1;
        }
        return 0;
    }
};

int main() {
    size_t n, m;
    cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    for (size_t i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        adj[x - 1].push_back(y - 1);
        adj[y - 1].push_back(x - 1);
    }
    int x, y;
    cin >> x >> y;
    Reachability reach = Reachability(adj);
    cout << reach.reach(x - 1, y - 1) << endl;
}
